### PR TITLE
[Sol->Yul] Refactor ByteArrayPop to use unchecked index access

### PR DESCRIPTION
**Pull request by mijovic**

### CODE REVIEW

File: libsolidity/codegen/YulUtilFunctions.h [hunk 1]
```diff
@@ -523,7 +523,7 @@ class YulUtilFunctions
 	/// @returns the name of a function that will clear given storage slot
 	/// starting with given offset until the end of the slot
 	/// signature: (slot, offset)
-	std::string partialClearStorageSlotFunction();
+	std::string partialClearStorageSlotFunction(bool _leftAlignedType);
```
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.h]: Maybe it would be better to use two different functions for the two alignments?

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: That is true, we can call explicitly the one we want instead of having parameter

Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.h]: Acutally the main purpose for the current function is to clear data in packed storage slots. If the alignment is different, then we should probably also use a different function name to clarify this distinction.

Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.h]: I'll do this later, not needed in this PR

File: libsolidity/codegen/YulUtilFunctions.h [hunk 2]
```diff
@@ -552,6 +552,11 @@ class YulUtilFunctions
 	/// signature: (data, len) -> data
 	std::string shortByteArrayEncodeUsedAreaSetLengthFunction();
 
+	/// @returns the name of a function that calculates slot and offset for index
+	/// Doesn't perform length checks, assumes that index is in bounds
+	/// signature: (array, index)
+	std::string longByteArrayStorageIndexAccessNoCheckFunction();
```
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.h:558]: I created this helper so I can use it in pop and as well in regular index access. Also can just have the same code twice if it is too much to create such a small function?

**Review by chriseth — APPROVED** [review]

File: test/cmdlineTests/name_simplifier/output [hunk 3]
```diff
@@ -139,14 +139,14 @@ object "C_59" {
             }
             function storage_array_index_access_struct_S_1252() -> slot, offset
             {
-                if iszero(lt(slot, 0x02)) { panic_error_0x32() }
-                slot := add(slot, slot)
+                if iszero(lt(offset, 0x02)) { panic_error_0x32() }
```
Comment (axic) [inline test/cmdlineTests/name_simplifier/output:142]: Random comment: why aren't these checks removed? `offset` is always 0.

Comment (hrkrshnn) [reply test/cmdlineTests/name_simplifier/output:142]: @axic This is a known issue. CommonSubExpression eliminator replaces `0` by return variables. And ExpressionSimplifier is not able to simplify this. There are two PRs proposing a fix https://github.com/ethereum/solidity/pull/11246/ and https://github.com/ethereum/solidity/pull/11243

**Pull request merged**